src/librustdoc/clean/types.rs

```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-crate struct Constant {
-    crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    /// Typed constant value.
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Generic { type_: Type, body: BodyId },
+    /// Inlined constant (from another crate).
+    Inline { type_: Type, did: DefId },
```

Comment:
Constants can be inlined without being from another crate. I think `Extern` or `ExternItem` would be a better name.

Reply:
Let's go for `Extern` then.